File: backend/app/services/supplier_service.py
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.models.supplier import Supplier
from app.schemas.supplier_schema import (
    SupplierCreate,
    SupplierUpdate,
)


def create_supplier(
    db: Session,
    supplier: SupplierCreate,
):
    existing = (
        db.query(Supplier)
        .filter(
            Supplier.email == supplier.email
        )
        .first()
    )

    if existing:
        raise ValueError(
            "Supplier with this email already exists."
        )

    new_supplier = Supplier(
        supplier_name=supplier.supplier_name,
        contact_person=supplier.contact_person,
        phone=supplier.phone,
        email=supplier.email,
        address=supplier.address,
    )

    try:
        db.add(new_supplier)
        db.commit()
        db.refresh(new_supplier)

    except IntegrityError:
        db.rollback()
        raise ValueError(
            "Unable to create supplier because of a database constraint."
        )

    return new_supplier


def get_all_suppliers(db: Session):
    return db.query(Supplier).all()


def get_supplier_by_id(
    db: Session,
    supplier_id: int,
):
    return (
        db.query(Supplier)
        .filter(
            Supplier.supplier_id == supplier_id
        )
        .first()
    )


def update_supplier(
    db: Session,
    supplier_id: int,
    supplier: SupplierUpdate,
):
    existing = get_supplier_by_id(
        db,
        supplier_id,
    )

    if not existing:
        raise ValueError(
            "Supplier not found."
        )

    duplicate = (
        db.query(Supplier)
        .filter(
            Supplier.email == supplier.email,
            Supplier.supplier_id != supplier_id,
        )
        .first()
    )

    if duplicate:
        raise ValueError(
            "Supplier with this email already exists."
        )

    existing.supplier_name = supplier.supplier_name
    existing.contact_person = supplier.contact_person
    existing.phone = supplier.phone
    existing.email = supplier.email
    existing.address = supplier.address

    try:
        db.commit()
        db.refresh(existing)

    except IntegrityError:
        db.rollback()
        raise ValueError(
            "Unable to update supplier because of a database constraint."
        )

    return existing


def delete_supplier(
    db: Session,
    supplier_id: int,
):
    supplier = get_supplier_by_id(
        db,
        supplier_id,
    )

    if not supplier:
        raise ValueError(
            "Supplier not found."
        )

    try:
        db.delete(supplier)
        db.commit()

    except IntegrityError:
        db.rollback()
        raise ValueError(
            "Cannot delete supplier because it is being used by another record."
        )

    return {
        "message": "Supplier deleted successfully."
    }

File: frontend/src/services/supplierService.ts
import api from "./api";

/**
 * --------------------------------------------------
 * Supplier Types
 * --------------------------------------------------
 */

export interface Supplier {
  supplier_id: number;
  supplier_name: string;
  contact_person: string;
  phone: string;
  email: string;
  address: string;
}

export interface SupplierCreate {
  supplier_name: string;
  contact_person: string;
  phone: string;
  email: string;
  address: string;
}

export interface SupplierUpdate {
  supplier_name: string;
  contact_person: string;
  phone: string;
  email: string;
  address: string;
}


/**
 * --------------------------------------------------
 * Get All Suppliers
 * --------------------------------------------------
 */
export const getSuppliers = async (): Promise<Supplier[]> => {
  const response = await api.get<Supplier[]>("/suppliers/");

  return response.data;
};


/**
 * --------------------------------------------------
 * Get Supplier By ID
 * --------------------------------------------------
 */
export const getSupplierById = async (
  supplierId: number
): Promise<Supplier> => {
  const response = await api.get<Supplier>(
    `/suppliers/${supplierId}`
  );

  return response.data;
};


/**
 * --------------------------------------------------
 * Create Supplier
 * --------------------------------------------------
 *
 * Admin and Manager only.
 */
export const createSupplier = async (
  supplier: SupplierCreate
): Promise<Supplier> => {
  const response = await api.post<Supplier>(
    "/suppliers/",
    supplier
  );

  return response.data;
};


/**
 * --------------------------------------------------
 * Update Supplier
 * --------------------------------------------------
 *
 * Admin and Manager only.
 */
export const updateSupplier = async (
  supplierId: number,
  supplier: SupplierUpdate
): Promise<Supplier> => {
  const response = await api.put<Supplier>(
    `/suppliers/${supplierId}`,
    supplier
  );

  return response.data;
};


/**
 * --------------------------------------------------
 * Delete Supplier
 * --------------------------------------------------
 *
 * Admin and Manager only.
 */
export const deleteSupplier = async (
  supplierId: number
): Promise<void> => {
  await api.delete(`/suppliers/${supplierId}`);
};